### ISSUE TITLE
Prompt enumerator to use save point from older form version

### ISSUE THREAD

**Issue by seadowg** [getodk/collect#4879]
~~Blocked by #5338~~

## Acceptance
- [x] Given I've filled out a form partially
When I restart the phone
And restart form entry for the same form
Then I am informed that I have work to recover from the time and date the form was being filled initially

- [x] Given I've filled out a form partially
When I restart the phone
And restart form entry for the same form
And click "Recover"
Then I start form entry in the hierarchy with my previous answers

- [x] Given I've filled out a form partially
When I restart the phone
And restart form entry for the same form
And click "Don't recover"
Then I start form entry on the first question with no prefilled questions

- [x] Given I've filled out a form partially
When I restart the phone
And update the form I'm filling
And restart form entry for the same form (the newer version)
Then I am informed that I have work to recover from the time and date the form was being filled initially

- [x] Given I've filled out a form partially
When I restart the phone
And update the form I'm filling
And restart form entry for the same form (the newer version)
And click "Recover"
Then I start form entry in the hierarchy with my previous answers
And the form I'm filling is the older version

- [x] Given I've filled out a form partially
When I restart the phone
And update the form I'm filling
And restart form entry for the same form (the newer version)
And click "Don't recover"
Then I start form entry on the first question with no prefilled questions
And the form I'm filling is the newer version

- [x] Given I've filled out a form partially
When I restart the phone
And update the form I'm filling
And restart form entry for the same form (the older version)
Then I am not informed that I have work to recover from the time and date the form was being filled initially
And the form I'm filling is the older version

## Notes

Problem that inspired this reported on the forum: https://forum.getodk.org/t/prevent-collect-to-reload-a-form-definition-while-editing-this-form/34079.

**Comment by lognaturel** [comment]
I initially understood that the update was happening during active form entry but I’m not as sure. 

Maybe it’s that if the form does not update, users can rely on savepoints. But if the form does update they’d have to show all versions of forms and open the older version to get their savepoint. There’s a broader question about what to do with form instances in the case of a newer form definition. Currently we only open instances with the exact version they were created with. However, that results in data loss in the case of a bad form definition being updated. Similarly, maybe savepoints should be by formId, not formid/version.

**Comment by seadowg** [comment]
> But if the form does update they’d have to show all versions of forms and open the older version to get their savepoint.

Ah do you think the user actually had to reopen the form? I'll ask that on the forum. From the description, they reopened Collect, and it returned to the form, but your version of events makes more sense.

> Currently we only open instances with the exact version they were created with. However, that results in data loss in the case of a bad form definition being updated. Similarly, maybe savepoints should be by formId, not formid/version.

We might just want to introduce a specific path for "savepoint for a different version" where we ask the user whether they want to keep that data (and fill out the old form) or get rid of it.

**Comment by lognaturel** [comment]
> do you think the user actually had to reopen the form?

Not explicitly. I think the FormEntryActivity went out of memory and reopened. If the same form version had been the latest, the savepoint would have loaded. But since the form version changed, there was no savepoint file.

I meant that the user could get their desired savepoint file to open but currently they can only do that by showing and then opening the deprecated version. 

> We might just want to introduce a specific path for "savepoint for a different version" where we ask the user whether they want to keep that data (and fill out the old form) or get rid of it.

That sounds great and is related to https://github.com/getodk/collect/issues/3494

In fact I think we should consider doing that with all instance edits.

**Comment by seadowg** [comment]
> Not explicitly. I think the FormEntryActivity went out of memory and reopened. If the same form version had been the latest, the savepoint would have loaded. But since the form version changed, there was no savepoint file.

I'm not sure if that makes sense. If `FormEntryActivity` is recreated it would load a form using the URI in its `Intent` which points at a specific form in the DB via the DB id not the form id. The updated form would be a different row in the DB right? I'm pretty sure we'd be loading the exact same version in the recreation case.

**Comment by lognaturel** [comment]
> If FormEntryActivity is recreated it would load a form using the URI in its Intent which points at a specific form in the DB via the DB id not the form id. 

Aha! So probably the whole app went out of memory and they relaunched the form? (That’s consistent with what the enumerator recalls). I think your general proposal of asking what form version to use is great. Making that understandable to users will be the challenge.

**Comment by seadowg** [comment]
> Aha! So probably the whole app went out of memory and they relaunched the form? (That’s consistent with what the enumerator recalls).

Yeah, and if that's the case then the app is behaving "correctly" but confusingly. 

> I think your general proposal of asking what form version to use is great.

It feels like it would make savepoints a little less confusing. More like other applications that ask "Do you want to recover your work?". I can write that up!

**Comment by seadowg** [comment]
@lognaturel I've sketched out acceptance for this. I'm currently thinking we should show a dialog with some text and the two actions ("Recover" and "Don't recover"). If you're happy with the flow here, I think we can just leave tweaking the text until we have a PR.

It feels like it might be nice to introduce this flow for all save points in the future (not just for when there is a save point for an older version). I think starting with this special case and then feeling out how well people understand it is a good step before that though.

**Comment by lognaturel** [comment]
The flow looks good and I'm ok with always opening the savepoint with the older form version (that may not always be the optimal thing to do but it's safe and consistent with what would have happened if the update came after that submission was made).

> tweaking the text until we have a PR.

Agreed though it may be helpful to know what dynamic information needs to be available. Ideally I think the message should include the date that the savepoint file was created. My first inclination was to also include the form version it was created for but because version is an arbitrary string that might not always be helpful. Maybe the savepoint file creation date is good enough. I'm not sure whether or not we should tell the user that the form version changed but I think that's a decision we can make later.

> And update the form I'm filling

You don't want to consider doing this immediately for any savepoint that may exist?

That said, I imagine the implementation will be somewhat different. For a savepoint matching the current form version, the savepoint file will be right there. For a savepoint for a past form version, it will be in a different folder and could be for any version of the form. It would be nice to be able to drive this off the filesystem so that there's only one source of truth but that might not be practical.

**Comment by seadowg** [comment]
> Agreed though it may be helpful to know what dynamic information needs to be available. Ideally I think the message should include the date that the savepoint file was created.

I was thinking we could just have static text like "...recover work from a previous version of the form". I do like the suggestion of including the date though, and I'd hope enumerators would find that helpful to understand what's going on. I'll add that to the issue description.

> You don't want to consider doing this immediately for any savepoint that may exist?

I'm not sure what you mean? This is meant to be the form being updated on the device to simulate the issue we saw (the new form version being opened).

**Comment by lognaturel** [comment]
> This is meant to be the form being updated on the device

We could provide this step any time there’s a savepoint file saved. If we do it only for the case where the form has updated, it will be an improvement, but one that is almost never experienced. And if someone does experience it but then gets the implicit savepoint flow another time, that will be confusing.

**Comment by seadowg** [comment]
> We could provide this step any time there’s a savepoint file saved.

Ah sorry I get you. I'd discussed that at https://github.com/getodk/collect/issues/4879#issuecomment-1373923751: 

> It feels like it might be nice to introduce this flow for all save points in the future (not just for when there is a save point for an older version). I think starting with this special case and then feeling out how well people understand it is a good step before that though.

I see your point about it being very rare though. I think we'd always need to do extra/different work to get this working with old version save points (as you point out in https://github.com/getodk/collect/issues/4879#issuecomment-1373968583), so I reckon we start here given we know it helps people out of a bad state (losing save points when a form is updated) and we can write up a separate issue for showing the prompt for all save points. Sound good?

**Comment by grzesiek2010** [comment]
I think we should decide how the recovery dialog should look like. In my draft pr it's:
![Screenshot_1706303177](https://github.com/getodk/collect/assets/3276264/dcf33361-d799-44b0-a470-77e11092079f)

@alyblenkin could you take a look? What do you think about it?

**Comment by alyblenkin** [comment]
@grzesiek2010  This is looking good. I like that we are including specific information with the date and time to help make it more informative so they can take action. 

My first thought is that we should add a title. Also, are we trying to say that recovering the version will be time-consuming or that it takes effort? If so, maybe we say something like this:

**Recover version?**
It could take some time to recover the version of this form saved on Fri, Jan 26, 2024 at 22:06. Do you want to recover your work?

Don't recover  | Recover

**Comment by grzesiek2010** [comment]
No, it won't be time-consuming. The goal is to give users a choice whether they want to recover their forms or not after updating a form.

**Comment by alyblenkin** [comment]
Ok sounds good. In that case, we can make the message more simple. 

**Recover version?**
The last version of this form was saved on Fri, Jan 26, 2024 at 22:06. Do you want to recover your work?

Don't recover | Recover


This might be a bit much, but I'm wondering if we should mention that it will restore the older version / revert back to the version saved on X date.

**Comment by lognaturel** [comment]
Here is the equivalent dialog in Enketo:

<img width="561" alt="Screenshot 2024-01-26 at 2 14 38 PM" src="https://github.com/getodk/collect/assets/967540/aa5bb923-b096-41f6-97ba-a52c4aed4d35">

Collect doesn't use "record" so it's not appropriate to align on but it's an interesting example.

Here's Excel's:

<img width="670" alt="Screenshot 2024-01-26 at 2 22 24 PM" src="https://github.com/getodk/collect/assets/967540/1d16125a-2c63-40a7-b788-f567a38ff992">

Here's an idea based on those:
```
Recover data?
Collect closed unexpectedly on <datetime> and saved in-progress data for this form.

Would you like to recover or discard it?
```

Or closer to what @alyblenkin proposed:

```
Recover your work?
Collect closed unexpectedly on <datetime> and saved your work.

Would you like to recover or discard it?
```

I'm not thrilled with either "in-progress data" or "your work", maybe one of you has a better suggestion! (Edit: the second one with "your work" is looking pretty good to me now)

I believe from the criteria that we decided not to say anything about the form version changing because it's confusing. I think we can treat that separately -- it's an issue that affects all edits of existing data. For example, you can have a draft created with version 4 of the form and when you open it it will use form version 4, even if form version 6 is the latest on device. Sometimes you'd really like to use version 6. From the acceptance criteria, we're matching that behavior: if a savepoint was saved for form version 4, even if form version 6 is the latest on device, the savepoint will be recovered for form version 4.

**Comment by alyblenkin** [comment]
Ah right make sense! I like the second one 

> **Recover your work?**
> Collect closed unexpectedly on <datetime> and saved your work.
> 
> Would you like to recover or discard it?

P.S. I don't know how you make the fancy callout, but it's very helpful!

**Comment by lognaturel** [comment]
> P.S. I don't know how you make the fancy callout, but it's very helpful!

I'm using a code block, it's a line with just three back ticks (```) before and after!

@grzesiek2010 I just had an annoying thought. Some users are using API form updates to support entity-like workflows. They may be producing a LOT of form versions. Do you have an order of magnitude sense of how long checking for savepoints for older versions takes? Like if you have 200 form versions on device, could it significantly affect form load time?

**Comment by grzesiek2010** [comment]
> @grzesiek2010 I just had an annoying thought. Some users are using API form updates to support entity-like workflows. They may be producing a LOT of form versions. Do you have an order of magnitude sense of how long checking for savepoints for older versions takes? Like if you have 200 form versions on device, could it significantly affect form load time?

It shouldn't be a problem. I've tested looping through 300 forms to find savepoints and it takes <0.3s (tested on Android 5 and 12)

**Comment by lognaturel** [comment]
This is VERY exciting, team! 🚀
